Form layout—Documentation (HDS-4886) (#2947)
hashicorp/design-system: website/docs/components/form/layout/partials/code/how-to-use.md

      <G.Legend>Create connection</G.Legend>
        <G.RadioCard @checked={{true}} {{on "change" this.onChange}} as |R|>
          <R.Icon @name="aws-color" />
          <R.Label>Quick peering with Quick Links</R.Label>
          <R.Badge @text="2-5 min" />
          <R.Description>Quick peering with quick links will provide the fastest way to connect to your providers’ network.</R.Description>
        </G.RadioCard>
        <G.RadioCard {{on "change" this.onChange}} as |R|>
          <R.Icon @name="aws-color" />
          <R.Label>Manual peering using AWS CLI</R.Label>
          <R.Badge @text="5-10 min" />
          <R.Description>Provide you AWS CLI template to apply connection settings.</R.Description>
        </G.RadioCard>
        <G.RadioCard {{on "change" this.onChange}} as |R|>
          <R.Icon @name="hcp" />
          <R.Label>Manual peering using HCP and AWS web console</R.Label>
          <R.Badge @text="30-60 min" />
          <R.Description>Manually follow UI instructions to complete configuring a connection at provider side.</R.Description>
        </G.RadioCard>
    </Hds::Form::RadioCard::Group>
  </FORM.Section>
</Hds::Form>
```

> ### Form Section Header, Title, & Description

Author: Added example code & explanatory text.

      </FG.Item>
    </FORM.SectionMultiFieldGroup>
  </FORM.Section>
</Hds::Form>
```

#### Responsive layout

In screen widths below 768px (the “md” breakpoint), the `SectionMultiFieldGroup` content layout will automatically stack.

```handlebars
<Hds::Form class="doc-form-layout-mobile-view" as |FORM|>
  <FORM.Section>
    <FORM.SectionMultiFieldGroup as |FG|>
      <Hds::Form::TextInput::Field as |F|>
        <F.Label>City</F.Label>
      </Hds::Form::TextInput::Field>

      <FG.Item @width="auto">
        <Hds::Form::Select::Field as |F|>
          <F.Label>State</F.Label>
          <F.Options>
            <option value="state-1">Ohio</option>
            <option value="state-2">Massachusetts</option>
            <option value="state-3">Washington</option>
            <option value="state-4">Florida</option>
            <option value="state-4">North Carolina</option>
          </F.Options>
        </Hds::Form::Select::Field>
      </FG.Item>

      <FG.Item @width="6em">
        <Hds::Form::TextInput::Field as |F|>
          <F.Label>Zip</F.Label>
        </Hds::Form::TextInput::Field>
      </FG.Item>
    </FORM.SectionMultiFieldGroup>
  </FORM.Section>
</Hds::Form>
```

> ### Form Footer

Author: Updated

The `FormFooter` should be used at the bottom of the other `Form` content to contain form actions for the form content. Use a `ButtonSet` to wrap 

```handlebars
<Hds::Form as |FORM|>
</Hds::Form>

> ```

Author: I added a section with an example form layout and another brief section showing how to override the default max-width for Sections & other content.

> ```

Reviewer: > I added a section with an example form layout

Nice! 

> another brief section showing how to override the default max-width for Sections & other content

Probably I would add a second code snippet, to show how to pass a custom value for, let's say, `36rem`. What do you think?

> ```

Author: Ok I can do that

> ```

Author: Added another example showing also how you can still individually override the max-width

  This “form” is actually just a div.
</Hds::Form>
```

### Form Header, Title, & Description

!!! Insight

The default Form Header Title `@tag` is `"div"` because the correct value is dependent on the individual page. We strongly encourage consumers to update the Title `@tag` to meet WCAG Success Criterion [1.3.1 Info and Relationships](https://www.w3.org/WAI/WCAG22/Understanding/info-and-relationships.html) as the visual experience should match what is presented to the user with assistive technology.

!!!

Use the optional `FormHeader` to include a `Title` and `Description` for your `Form`.

```handlebars
<Hds::Form as |FORM|>
  <FORM.Header>
    <FORM.HeaderTitle @tag="h2">My form title</FORM.HeaderTitle>
    <FORM.HeaderDescription>A brief description of my form content.</FORM.HeaderDescription>
  </FORM.Header>
</Hds::Form>
```

#### Form Header Title tag & size

The `@tag` argument changes the HTML element that wraps the `FormHeaderTitle` content. When organizing the content on a webpage, the heading levels should reflect the structure of the page. For example, if a `FormHeaderTitle` appears directly below the main heading of the page, it should be `"h2"`.

> To specify which size the `FormHeaderTitle` displays at, use the `@tag` argument.

Reviewer: ```suggestion
To specify which size the `FormHeaderTitle` displays at, use the `@size` argument.
```

Author: Pushed up fix

The default Form Section Header Title `@tag` is `"div"` because the correct value is dependent on the individual page. We strongly encourage consumers to update the Title `@tag` to meet WCAG Success Criterion [1.3.1 Info and Relationships](https://www.w3.org/WAI/WCAG22/Understanding/info-and-relationships.html) as the visual experience should match what is presented to the user with assistive technology.

!!!

Similarly to the `Form`, each `FormSection` can optionally include its own `SectionHeader` with `Title` and `Description`.

```handlebars
<Hds::Form as |FORM|>
  <FORM.Section>
    <FORM.SectionHeader>
      <FORM.SectionHeaderTitle @tag="h3">Section header title</FORM.SectionHeaderTitle>
      <FORM.SectionHeaderDescription>
        Section Header description
      </FORM.SectionHeaderDescription>
    </FORM.SectionHeader>
  </FORM.Section>
</Hds::Form>
```

#### Section Header Title tag & size

The `@tag` argument changes the HTML element that wraps the `SectionHeaderTitle` content. When organizing the content on a webpage, the heading levels should reflect the structure of the page. For example, if the `FormHeaderTitle` tag is `"h3"`, the Section Title should be `"h4"`.

> To specify which size the `SectionHeaderTitle` displays at, use the `@tag` argument.

Reviewer: ```suggestion
To specify which size the `SectionHeaderTitle` displays at, use the `@size` argument.
```

Author: Fixed

## How to use this component

Use the `Form` to contain other Form layout components and form content. This establishes consistent spacing for each `FormSection` and the other content.

### Form

```handlebars{data-execute=false}
<Hds::Form>
  Add form content here.
</Hds::Form>
```

The `Form` renders as an HTML `form` element by default. Use the `tag` argument to optionally use an HTML `div` instead.

```handlebars{data-execute=false}
<Hds::Form @tag="div">
  This “form” is actually just a div.
</Hds::Form>
```

### Form Header, Title, & Description

!!! Insight

The default Form Header Title `@tag` is `"div"` because the correct value is dependent on the individual page. We strongly encourage consumers to update the Title `@tag` to meet WCAG Success Criterion [1.3.1 Info and Relationships](https://www.w3.org/WAI/WCAG22/Understanding/info-and-relationships.html) as the visual experience should match what is presented to the user with assistive technology.

!!!

> Use the optional `FormHeader` to include a `Title` and `Description` for your `Form`.

Author: @didoo Do you have any opinions to add?

> Use the optional `FormHeader` to include a `Title` and `Description` for your `Form`.

Reviewer: Not strong ones. Whatever you decide, works for me (there's not an easy clear cut solution to this repetitions, across the documentation)

> Use the optional `FormHeader` to include a `Title` and `Description` for your `Form`.

Reviewer 2: @KristinLBradley I'm good with whatever you think is best as well. Good to leave it as is.

</Hds::Form>
```

To control the widths of individual elements within a `SectionMultiFieldGroup`, you can wrap the element with an `Item` and pass in a `width` value. Fields not wrapped with an `Item` will take up the remaining available width.

```handlebars
<Hds::Form as |FORM|>
  <FORM.Section>
    <FORM.SectionMultiFieldGroup as |FG|>
      <Hds::Form::TextInput::Field as |F|>
        <F.Label>City</F.Label>
      </Hds::Form::TextInput::Field>

      <FG.Item @width="auto">
        <Hds::Form::Select::Field as |F|>
          <F.Label>State</F.Label>
          <F.Options>
            <option value="state-1">Ohio</option>
            <option value="state-2">Massachusetts</option>
            <option value="state-3">Washington</option>
            <option value="state-4">Florida</option>
            <option value="state-4">North Carolina</option>
          </F.Options>
        </Hds::Form::Select::Field>
      </FG.Item>

      <FG.Item @width="8em">
        <Hds::Form::TextInput::Field as |F|>
          <F.Label>Zip</F.Label>
        </Hds::Form::TextInput::Field>
      </FG.Item>
    </FORM.SectionMultiFieldGroup>
  </FORM.Section>
</Hds::Form>
```

#### Responsive layout

> In screen widths below 768px (the “md” breakpoint), the `SectionMultiFieldGroup` content layout will automatically stack.

Reviewer 2: ```suggestion
In screen widths below 768px (the [“md” breakpoint](/foundations/breakpoints)), the `SectionMultiFieldGroup` content layout will automatically stack.
```
[Nit] Could align to the guidlines and add a link for the breakpoint.

```handlebars
<Hds::Form class="doc-form-layout-mobile-view" as |FORM|>
  <FORM.Section>
    <FORM.SectionMultiFieldGroup as |FG|>
      <Hds::Form::TextInput::Field as |F|>
        <F.Label>City</F.Label>
      </Hds::Form::TextInput::Field>

      <FG.Item @width="auto">
        <Hds::Form::Select::Field as |F|>
          <F.Label>State</F.Label>
          <F.Options>
            <option value="state-1">Ohio</option>
            <option value="state-2">Massachusetts</option>
            <option value="state-3">Washington</option>
            <option value="state-4">Florida</option>
            <option value="state-4">North Carolina</option>
          </F.Options>
        </Hds::Form::Select::Field>
      </FG.Item>

      <FG.Item @width="6em">
        <Hds::Form::TextInput::Field as |F|>
          <F.Label>Zip</F.Label>
        </Hds::Form::TextInput::Field>
      </FG.Item>
    </FORM.SectionMultiFieldGroup>
  </FORM.Section>
</Hds::Form>
```

### Form Footer

> The `FormFooter` should be used at the bottom of the other `Form` content to contain form actions. Use a [Button Set](/components/button-set) to wrap and set spacing for Buttons.

Reviewer 2: ```suggestion
The `FormFooter` should be used at the bottom of the other `Form` content to contain form actions. Use the yielded `ButtonSet` to wrap and set spacing for Buttons.
```
[Suggestion] Since the button set comes from the footer itself, and they don't need to use an `Hds::ButtonSet` directly, may be more clear to not link out to those docs, or at lease use the yielded component name format.
